logging/README.md

```diff
@@ -0,0 +1,228 @@
+        super.setLoggingConfiguration(config);
+        super.setErrorManager(error);
+    }
+
+    @Override
+    public Map<String, ?> format(ExtLogRecord record, TraceInfo trace) {
+        return super.toEsc(record, trace);
+    }
+
+    @Override
+    protected boolean shouldIncludeParameter(Object p) {
+        return !(p instanceof MyCustomParameterType);
+    }
+}
+```
+
+### Resource type and labels
+You can configure the resource type and the resource labels. This is used by GCP operations as dimensions
+when analysing logs. Please refer to the [GCP documentation](https://cloud.google.com/logging/docs/api/v2/resource-list#resource-types) 
+for the allowable resource types and their associated labels. 
+
+* `quarkus.google.cloud.logging.resource.type` - The type string, defaults to `generic_node`
+* `quarkus.google.cloud.logging.resource.label.<string>` - This is a string map with key/value pairs
+
+### Log record labels
+You can configure default labels for each log record:
+
+* `quarkus.google.cloud.logging.default-label.<string>` - This is a string map with key/value pairs
+
+You can also extract custom labels per record by binding a `LabelExtractor` to the CDI context, e.g.: 
+
+```java
+package mypackage;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.logmanager.ExtLogRecord;
+
+import io.quarkiverse.googlecloudservices.logging.runtime.LabelExtractor;
+
+@ApplicationScoped
+public class TestLabelExtractor implements LabelExtractor {
+
+    public Map<String, String> extract(ExtLogRecord record) {
+        if (record.getParameters() == null) {
+            return null;
+        } else {
+            // do something smart here
+        }
+    }
+}
+```
+
+### Tracing
+In order to include trace and span ID in the log you must bind a `TraceInfoExtractor` to the CDI context, e.g.:
+
+```java
+package mypackage;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.logmanager.ExtLogRecord;
+
+import io.quarkiverse.googlecloudservices.logging.runtime.TraceInfo;
+import io.quarkiverse.googlecloudservices.logging.runtime.TraceInfoExtractor;
+
+@ApplicationScoped
+public class TestTraceInfoExtractor implements TraceInfoExtractor {
+
+    @Override
+    public TraceInfo extract(ExtLogRecord record) {
+        return new TraceInfo(/* something, something */);
+    }
+}
+```
+
+This will include the trace and span ID in the ESC logging if they are not null. By default this
+will also be set on the GCP operations log entry as a link to GCP operations tracing, but this is
+configurable: 
+
+* `quarkus.google.cloud.logging.gcp-tracing.enabled=[true|false]`
+
+# Injecting GCP Logging
+You can inject a `Logging` instance directly. If you do, the configuration for the project to use,
+still apply. 
+
+If you want the configuration for default log, resource type and labels, and default labels, you can
+inject a `WriteOptionsHolder` which contains an array of default write options as configured. 
```

Comment:
I don't really understand the WriteOptions stuff, can you explain it to me ?

Reply:
1 - If you want to inject a `com.google.cloud.logging.Logging` to use directly (instead of logging normally to `jboss` or `slf4j` etc), you will need the configured write options. 

2 - The write options are an array, so for simplicity of injection I opted to wrap the array in a provider, and opted to call it a `*Holder` instead since to align the name with `GcpConfigHolder gcpConfigHolder`

Reply:
OK, understood.